Allow vacuum.send_command to return responses

SLaks opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This allows scripts to read information about
vacuums that is not exposed by the integration,
such as WiFi signal strength.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

gjohansson-ST reviewed (changes requested):
As @Lash-L already commented this needs to be split up. Thanks 👍

SLaks committed:
Allow vacuum.send_command to return responses

This allows scripts to read information about
vacuums that is not exposed by the integration,
such as WiFi signal strength.

SLaks commented:
Pushed to remove the integration-specific changes.

SLaks commented:
I also pushed two more branches for the two integrations; should I send their PRs now or after this is merged?

Lash-L commented:
> I also pushed two more branches for the two integrations; should I send their PRs now or after this is merged?

Personally I'd wait. You never know what changes you'll have to make here. You also don't want your other PRs to end up getting buried. 

I'd fix the ci on this PR and then once you get this one merged in, make the others

SLaks commented:
Fixed

gjohansson-ST commented:
Putting this back into draft as test coverage should be added, not removing the need off.

Lash-L commented:
> The coverage error was from https://github.com/home-assistant/core/commit/346a7cdd2eea815bcf2ca291b6e669ca0c59dce9#annotation_18915759351
> 
> 
> 
> How do I cover that? https://github.com/home-assistant/core/blob/dev/tests/components/vacuum/test_init.py has no tests for anything like this code, which is why I skipped coverage.

I think the send command is originally covered by demo. So go in tests/components/demo/vacuum and look there

gjohansson-ST commented:
We should add coverage in it's own tests I think and not to be covered by `demo` or others.
Take a look at other entity components which also implements calls with return values such as `weather` and "copy" the idea from how it's done there.

I guess it's a bit unfortunate that `vacuum` is missing out here on more testing.

SLaks commented:
That test [mocks the service entirely](https://github.com/home-assistant/core/blob/dev/tests/components/demo/test_vacuum.py#L191), so it can't cover this.

And, the [demo vacuum](https://github.com/home-assistant/core/blob/dev/homeassistant/components/demo/vacuum.py#L177) overrides `async_send_command()`, so it cannot be used to cover the base implementation anyway.

Lash-L commented:
I've expanded vacuum testing here:
https://github.com/home-assistant/core/pull/112523

This should fix any problems you have with coverage

edenhaus commented:
We discussed this PR internally and concluded that we don't accept it as for using the `send_command` service call a user needs to the docs and check what's supported and how it can be used. Instead of using the `send_command`, integrations should provide their custom services so that the UI can show and suggest the correct data for the service, including translation.
A custom service is more user-friendly than a generic one, where the user needs to know the exact data format and all the possible values.

In the next weeks, I will analyze all vacuum integration and evaluate if some common services should be added to the base entity.

> This allows scripts to read information about
vacuums that is not exposed by the integration,
such as WiFi signal strength.

In this case, it's better to add a sensor entity for the wifi signal instead of each user creating a template sensor independently.
We should improve the integration and add missing parts instead of using a generic service. As I'm maintaining the Ecovacs integration, I understand that this generic service is used to issue commands that are not yet supported. (Also, the Ecovacs integration has support for it, which I will deprecate soon and replace the common use case with custom services)

Someone will think now that without the generic service, I need to wait for the integration to support my custom command, but at the point where the user knows which command and values are required to send to the bot, he could also create a PR/issue for the missing command. After the command is added, all users will benefit from it, not just a single one.

SLaks commented:
Then why does `send_command` exist at all?

The is most useful for exploring commands or or debugging vacuum integrations, or for command responses that aren't worth adding as sensors (eg, WiFi bssid, internal firmware versions), but which advanced users may want to play with.

frenck commented:
> Then why does send_command exist at all?

History. It is one of the things we wouldn't allow to be introduced again these days.

> The is most useful for exploring commands or or debugging vacuum integrations, or for command responses that aren't worth adding as sensors (eg, WiFi bssid, internal firmware versions), but which advanced users may want to play with.

That is not a good use case. If there is a functionality that can be provided by an integration, it should provide it, in a controlled matter. Home Assistant isn't a Postman debugging tool 😉 

../Frenck

SLaks commented:
Makes sense

Lash-L commented:
> In the next weeks, I will analyze all vacuum integration and evaluate if some common services should be added to the base entity. Please feel free to help and suggest/share ideas. (One is the floorplan but I need to analyze it how we can provide this information to HA, including rooms)

As Roborock codeowner, I'm happy to help in anyway possible with this - even if it is just a second set of eyes or hands. I've put a decent bit of thought into this in a few different arch discussions.

> Someone will think now that without the generic service, I need to wait for the integration to support my custom command, but at the point where the user knows which command and values are required to send to the bot, he could also create a PR/issue for the missing command. After the command is added, all users will benefit from it, not just a single one.

I agree with you conceptually. Send command is definitely worse than an actual entity. However, looking at ecovacs implementation it might be slightly different than my Roborock implementation and I think that is worth bringing up.

send command is used to send one of the 222 roborock commands( that I currently know of - more are added with each device). They are sent directly to the mqtt broker, I don't do any kind of processing on them. This allows users to use complete functionality that I don't even know about or that is provided with a device I can't test and is challenging for me to add functionality for and I have never had a user attempt to add functionality to the roborock integration before @SLaks so I'm not sure if counting on end users to add the support instead of send command is likely.

As well, I started the process for the ability to clean rooms in May of 2023, but I had to pause and wait for real support to be able to add it, so in the mean time I explained how to do it via the send_command service. Without it, for the past 9 months, users would have been unable to clean any rooms from the roborock integration and that is a lot of users primary use case of the integration.

All of this to say, I think send_command is super helpful, I agree that entities are better when possible, but I just wanted to say this now to try to make sure the discussion never went towards a more general depreciation (rather than just in ecovacs) of send_command

Hearing you say it all, I do think adding the responses to send_command is probably not the best as it furthers the reliance end users may have on the service.
